hedgehog/src/Hedgehog/Internal/Property.hs

```diff
@@ -317,6 +319,10 @@ instance Monad m => Monad (TestT m) where
       unTest m >>=
       unTest . k
 
+  fail =
+    Fail.fail
+
+instance Monad m => MonadFail (TestT m) where
   fail err =
```

Comment:
We need one for `PropertyT` as well

```haskell
instance MonadPlus m => MonadFail (PropertyT m) where
  fail _ =
    mzero
```

---

Here's an example:

```haskell
maybeOne :: Gen (Maybe Int)
maybeOne = do
  Just n <-
    pure <$> Gen.constant (Just 1)
  return n

prop_maybeOne :: Property
prop_maybeOne =
  property $ do
    Just x <- forAll maybeOne
    assert $
      x == 1
```

This doesn't compile without a `MonadFail (PropertyT m)` instance:

```
        * No instance for (Control.Monad.Fail.MonadFail (PropertyT IO))
            arising from a do statement
            with the failable pattern `Just x'
        * In a stmt of a 'do' block: Just x <- forAll maybeOne
          In the second argument of `($)', namely
            `do Just x <- forAll maybeOne
                assert $ x == 1'
          In the expression:
            property
              $ do Just x <- forAll maybeOne
                   assert $ x == 1
       |
    39 |     Just x <- forAll maybeOne
       |     ^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
I considered that as well but thought the semantics wouldn't be very clear. For a generator, `mzero` and `fail` being discard is pretty clear. While for a test, `fail` being a concrete failure is pretty clear. For a property... I don't think there's a single correct answer, so I omitted the instance.

Reply:
Yeah, that's true. Better be explicit and let the compiler do its job, then.

Reply:
However, now that we've agreed to go with `fail = error` until the next major version bump, it might be worth considering

```haskell
instance MonadPlus m => MonadFail (PropertyT m) where
  fail =
    error
```

here as well. ━ Just we don't break any builds for those switching to GHC 8.6.

Reply:
People expect to see MonadFail errors when upgrading to 8.6. It's part and parcel with the upgrade and that's a good thing. We shouldn't add a `MonadFail` instances just to counter that.

Reply:
Yeah, I think `fail` for a generator and a test is fair enough.

Reply:
/cc @fosskers